lib/cloudkeeper/one/core_connector.rb

```diff
@@ -1,36 +1,84 @@
+      end
+
       def pre_action(_empty, _unused_call)
-        raise NotImplementedError, 'this call is not implemented yet'
+        logger.debug 'Running \'pre-action\'...'
+        handle_errors { remove_expired }
       end
 
       def post_action(_empty, _unused_call)
-        raise NotImplementedError, 'this call is not implemented yet'
+        logger.debug 'Running \'post-action\'...'
+        CloudkeeperGrpc::Status.new(code: :SUCCESS, message: '')
       end
 
-      def add_appliance(_appliance, _unused_call)
-        raise NotImplementedError, 'this call is not implemented yet'
+      def add_appliance(appliance, _unused_call)
+        logger.debug "Registering appliance #{appliance.identifier.inspect}"
+        handle_errors { register_or_update_appliance appliance }
       end
 
-      def update_appliance(_empty, _unused_call)
-        raise NotImplementedError, 'this call is not implemented yet'
+      def update_appliance(appliance, _unused_call)
+        logger.debug "Updating appliance #{appliance.identifier.inspect}"
+        handle_errors { appliance.image ? register_or_update_appliance(appliance) : update_appliance_metadata(appliance) }
       end
 
-      def remove_appliance(_empty, _unused_call)
-        raise NotImplementedError, 'this call is not implemented yet'
+      def remove_appliance(appliance, _unused_call)
+        logger.debug "Removing appliance #{appliance.identifier.inspect}"
+        handle_errors { remove_appliance appliance }
       end
 
-      def remove_image_list(_empty, _unused_call)
-        raise NotImplementedError, 'this call is not implemented yet'
+      def remove_image_list(image_list_identifier, _unused_call)
+        logger.debug "Removing appliances from image list #{image_list_identifier.image_list_identifier.inspect}"
+        handle_errors { remove_image_list image_list_identifier.image_list_identifier }
       end
 
       def image_lists(_empty, _unused_call)
-        raise NotImplementedError, 'this call is not implemented yet'
+        logger.debug 'Retrieving image lists registered in OpenNebula'
+        list_image_lists.each
+      end
+
+      def appliances(image_list_identifier, _unused_call)
+        logger.debug "Retrieving appliances from image list #{image_list_identifier.image_list_identifier.inspect} " \
+                     'registered in OpenNebula'
+        list_appliances(image_list_identifier.image_list_identifier).each
       end
 
-      def appliances(_empty, _unused_call)
-        raise NotImplementedError, 'this call is not implemented yet'
+      private
+
+      def handle_errors
+        raise Cloudkeeper::One::Errors::ArgumentError, 'Error handler was called without a block!' unless block_given?
+
+        yield
+      rescue Cloudkeeper::One::Errors::StandardError => ex
+        CloudkeeperGrpc::Status.new(code: ERRORS[ex.class], message: ex.message)
+      else
+        CloudkeeperGrpc::Status.new(code: :SUCCESS, message: '')
```

Comment:
I wouldn't do this. It will transform any unexpected error (not a subclass of `Cloudkeeper::One::Errors::StandardError`) into a SUCCESS. Even letting the call crash without a valid response would be more acceptable.